### PR TITLE
[jit] Add support for `@staticmethod`

### LINKED ISSUE #26792: torch.jit.script cannot compile forward() calling a @staticmethod annotated method
## 🐛 Bug

The following code could reproduce the problem:

    import torch.nn
    import torch.jit
        
    class Bug(torch.nn.Module):
        @staticmethod
        def foo():
            pass
    
        def forward(self):
            Bug.foo()
            return torch.tensor([0])
    
    
    bug = Bug()
    torch.jit.script(bug)

```
Traceback (most recent call last):
  File "bug.py", line 15, in <module>
    torch.jit.script(bug)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/__init__.py", line 1162, in script
    return _convert_to_script_module(obj)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/__init__.py", line 1812, in _convert_to_script_module
    return WeakScriptModuleProxy(mod, stubs)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/__init__.py", line 1386, in init_then_register
    original_init(self, *args, **kwargs)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/__init__.py", line 1736, in __init__
    _create_methods_from_stubs(self, stubs)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/__init__.py", line 1347, in _create_methods_from_stubs
    self._c._create_methods(self, defs, rcbs, defaults)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/__init__.py", line 1058, in _compile_and_register_class
    ast = get_jit_class_def(obj, obj.__name__)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/frontend.py", line 143, in get_jit_class_def
    self_name=self_name) for method in methods]
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/frontend.py", line 143, in <listcomp>
    self_name=self_name) for method in methods]
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/frontend.py", line 166, in get_jit_def
    return build_def(ctx, py_ast.body[0], type_line, self_name)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/frontend.py", line 195, in build_def
    param_list = build_param_list(ctx, py_def.args, self_name)
  File "/home/d00413986/.local/share/virtualenvs/Code-ekJV9_xh/lib/python3.6/site-packages/torch/jit/frontend.py", line 215, in build_param_list
    raise ValueError(_vararg_kwarg_err)
ValueError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults
```
## Expected behavior
Code compiles

## Environment

PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: TITAN Xp
GPU 1: TITAN Xp
GPU 2: TITAN Xp
GPU 3: TITAN Xp

Nvidia driver version: 418.43
cuDNN version: /usr/local/cuda-10.0/lib64/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] pytorch-transformers==1.2.0
[pip3] torch==1.2.0
[conda] Could not collect

## Additional context

cc @suo

### PR DESCRIPTION
**Pull request by driazati**
Resolve static methods as functions

Fixes #26792

Differential Revision: [D17695094](https://our.internmc.facebook.com/intern/diff/17695094/)

### CODE REVIEW

**Comment by eellison** [comment]
Going forward I'm not sure if we should implement a feature for `nn.Modules` without also implementing it for User Defined Classes.

**Comment by driazati** [comment]
@eellison it should work for class types too, added a test

**Comment by eellison** [comment]
Could you add a tests for when two different classes define a static method with the same name, and those two methods are used in the same function ?

File: test/test_jit.py [hunk 1]
```diff
@@ -3544,6 +3544,33 @@ def bar(x):
...
+        class N(nn.Module):
+            def __init__(self):
+                super(N, self).__init__()
+
+            @staticmethod
+            def my_method(x):
+                return x * 100
+
+            def forward(self, x):
+                return x - M.my_method(x) + N.my_method(x)
```
Comment (eellison) [inline test/test_jit.py]: How does this serialize if we're not inlining methods ? Could you print out the source ?

**Review by eellison — COMMENTED** [review]
Looks good i just have a couple questions / nits

File: test/test_jit.py (hunk 1 above)
Comment (driazati) [reply test/test_jit.py]: We still do inlining:

```
graph(%self : ClassType<N>,
      %x.1 : Tensor):
  %23 : int = prim::Constant[value=100]() # test/test_jit.py:3654:27
  %22 : int = prim::Constant[value=1]()
  %24 : Tensor = aten::add(%x.1, %23, %22) # test/test_jit.py:3654:23
  %7 : Tensor = aten::sub(%x.1, %24, %22) # test/test_jit.py:3668:23
  %26 : Tensor = aten::mul(%x.1, %23) # test/test_jit.py:3665:23
  %12 : Tensor = aten::add(%7, %26, %22) # test/test_jit.py:3668:23
  return (%12)
```

With inlining off the methods get correctly mangled:

```
graph(%self : ClassType<N>,
      %x.1 : Tensor):
  %9 : Function = prim::Constant[name="my_method"]()
  %6 : int = prim::Constant[value=1]()
  %4 : Function = prim::Constant[name="my_method"]()
  %5 : Tensor = prim::CallFunction(%4, %x.1) # test/test_jit.py:3668:27
  %7 : Tensor = aten::sub(%x.1, %5, %6) # test/test_jit.py:3668:23
  %10 : Tensor = prim::CallFunction(%9, %x.1) # test/test_jit.py:3668:44
  %12 : Tensor = aten::add(%7, %10, %6) # test/test_jit.py:3668:23
  return (%12)
```

```
Archive:  a.pt
 extracting: a/version               
 extracting: a/data.pkl              
  inflating: a/code/__torch__.py     
  inflating: a/code/__torch__.py.debug_pkl  
  inflating: a/code/__torch__/___torch_mangle_0.py  
  inflating: a/code/__torch__/___torch_mangle_0.py.debug_pkl  
  inflating: a/code/__torch__/___torch_mangle_1.py  
  inflating: a/code/__torch__/___torch_mangle_1.py.debug_pkl  
 extracting: a/constants.pkl         
[dev] ~/d/pytorch > cat ^C
[dev] ~/d/pytorch > cd ^C
[dev] ~/d/pytorch > cat a/code/__torch__/___torch_mangle_0.py   (base) [ 15:22:38 ]
op_version_set = 1
def my_method(x: Tensor) -> Tensor:
  return torch.add(x, 100, 1)
[dev] ~/d/pytorch > cat a/code/__torch__/___torch_mangle_1.py   (base) [ 15:22:50 ]
op_version_set = 1
def my_method(x: Tensor) -> Tensor:
  return torch.mul(x, 100)
[dev] ~/d/pytorch >                                             (base) [ 15:22:53 ]
```

and the tests still work

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -499,6 +497,14 @@ struct Environment {
       retval = resolver->resolveValue(ident, method, range);
     }
 
+    if (!retval) {
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: Still dont really understand why this was changed

**Review by eellison — APPROVED** [review]
Looks good

File: torch/csrc/jit/script/compiler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (driazati) [reply torch/csrc/jit/script/compiler.cpp]: Static methods aren't put on to class types but resolved into functions directly from Python, so if the class is resolved as a type then it won't find the static methods

**Pull request closed without merging**